CASSANDRA-20081: Add compaction type and strategy properties to compaction history

arvindKandpal-ksolves opened the pull request:
Currently, `nodetool compactionhistory` does not provide visibility into the specific type of compaction (e.g., Major, Minor, Validation) or the compaction strategy used (e.g., STCS, LCS). Additionally, the output displays raw bytes by default, which is difficult to read.

This patch enhances the compaction history by persisting the compaction type and strategy properties into the `system.compaction_history` table and exposing them via `nodetool`.

**Changes included in this patch:**
1.  **Schema Update:** Added `compaction_type` column to `system.compaction_history`.
2.  **Logic Update:** Updated `CompactionTask` to capture the compaction type (e.g., Major, Cleanup) and strategy class (e.g., SizeTieredCompactionStrategy) and persist them.
3.  **JMX/Tooling:** Updated `CompactionHistoryTabularData` and `nodetool compactionhistory` printer to display these new fields.
4.  **UX Improvement:** Changed `nodetool compactionhistory` to display human-readable sizes (KiB, MiB) by default. Added a `-n` / `--no-human-readable` flag to revert to raw bytes.
5.  **Tests:** Updated `SystemKeyspaceTest`, `CompactionTaskTest`, and `CompactionHistoryTest` to cover schema changes and verify the persistence of new fields.

**Tests Performed:**
* `ant test -Dtest.name=SystemKeyspaceTest` (PASSED)
* `ant test -Dtest.name=CompactionTaskTest` (PASSED)
* `ant test -Dtest.name=CompactionHistoryTest` (PASSED)
* Manual verification using `nodetool compactionhistory` on a local node.

patch by arvindKandpal-ksolves ;  for CASSANDRA-20081

arvindKandpal-ksolves committed:
CASSANDRA-20081: Add compaction type and strategy properties to compaction history

jyothsnakonisa commented on test/unit/org/apache/cassandra/db/compaction/CompactionTaskTest.java:
```diff
@@ -71,7 +71,7 @@ public void setUp() throws Exception
-    @Test
+@Test
     public void testTaskIdIsPersistedInCompactionHistory()
     {
```
Please fix the indentation
```suggestion
    @Test
    public void testTaskIdIsPersistedInCompactionHistory()
    {
```

jyothsnakonisa commented on test/unit/org/apache/cassandra/db/compaction/CompactionTaskTest.java:
```diff
@@ -103,6 +103,12 @@ public void testTaskIdIsPersistedInCompactionHistory()
+        String type = one.getString("compaction_type");
+        Assert.assertEquals("Compaction", type);
+
+        java.util.Map<String, String> properties = one.getMap("compaction_properties", org.apache.cassandra.db.marshal.UTF8Type.instance, org.apache.cassandra.db.marshal.UTF8Type.instance);
```
Please avoid fully qualified class names
```suggestion
        Map<String, String> properties = one.getMap("compaction_properties", UTF8Type.instance, UTF8Type.instance);
```

jyothsnakonisa commented on src/java/org/apache/cassandra/db/compaction/CompactionHistoryTabularData.java:
```diff
@@ -30,11 +30,13 @@
     private static final String[] ITEM_NAMES = new String[]{ "id", "keyspace_name", "columnfamily_name", "compacted_at",
-                                                             "bytes_in", "bytes_out", "rows_merged", "compaction_properties" };
+                                                             "bytes_in", "bytes_out", "rows_merged", "compaction_properties",
+                                                             "compaction_type" }; // Added
```
Comment `// Added` might not be needed

jyothsnakonisa commented on src/java/org/apache/cassandra/db/compaction/CompactionHistoryTabularData.java:
```diff
@@ -30,11 +30,13 @@
                                                              "column family name", "compaction finished at",
-                                                             "total bytes in", "total bytes out", "total rows merged", "compaction properties" };
+                                                             "total bytes in", "total bytes out", "total rows merged",
+                                                             "compaction properties", "compaction type" }; // Added
```
Comment `// Added` might not be needed

jyothsnakonisa commented on src/java/org/apache/cassandra/db/compaction/CompactionHistoryTabularData.java:
```diff
@@ -45,15 +47,16 @@ public class CompactionHistoryTabularData
-    
+
     public static final String COMPACTION_TYPE_PROPERTY = "compaction_type";
-    
-    static 
+
```
Since Compaction Type property is another column in the table, you can get rid of this constant.
Also, noticed that `CompactionHistorySystemTableUpgradeTest` has compilation errors with these changes that you introduced in this patch.

Also, remove unused imports in files where this constant is used.

jyothsnakonisa commented on src/java/org/apache/cassandra/db/compaction/CompactionTask.java:
Please remove unused imports in this file

jyothsnakonisa commented on test/unit/org/apache/cassandra/db/SystemKeyspaceTest.java:
```diff
@@ -158,6 +158,34 @@ public void testPersistLocalMetadata()
+        String compactionType = "TestMajor";
+
+        SystemKeyspace.updateCompactionHistory(
+            org.apache.cassandra.utils.TimeUUID.Generator.nextTimeUUID(),
```
```suggestion
            TimeUUID.Generator.nextTimeUUID(),
```

jyothsnakonisa commented on src/java/org/apache/cassandra/db/SystemKeyspace.java:
Can you please check `SystemKeyspaceMigrator41.java`?  specifically `migrateCompactionHistory()` method. You might need to include compaction_type during migration.

jyothsnakonisa commented on src/java/org/apache/cassandra/tools/nodetool/CompactionHistory.java:
```diff
@@ -33,10 +33,10 @@ public class CompactionHistory extends AbstractCommand
+    @Option(paramLabel = "no_human_readable",
+            names = { "-n", "--no-human-readable" },
+            description = "Display raw bytes (disable default human readable format)")
+    private boolean noHumanReadable = false;
```
I am concerned about changing the flag name and default behavior. Imagine some scripts using this node tool command there is a chance of breaking them.

jyothsnakonisa reviewed (commented):
Left few comments.

arvindKandpal-ksolves commented:
@jyothsnakonisa Thanks for the review. I have addressed all your comments:
1. Reverted nodetool flag changes to maintain backward compatibility.
2. Updated SystemKeyspaceMigrator41 and fixed upgrade tests.
3. Fixed code style, imports, and cleanup issues.
4. Ready for re-review.

jyothsnakonisa reviewed (approved):
@arvindKandpal-ksolves  looks good to me mostly. Can you also run CI and update the report on the JIRA

arvindKandpal-ksolves commented:
Hi @jyothsnakonisa , thanks for the approval!

Could you please tell me how I can trigger the CI and what need to update in JIRA , as I am contributing to the code segment in Cassandra first time ?

jyothsnakonisa commented:
Hi @arvindKandpal-ksolves @smiklosovic might be able to help you with that. I generally run it on my organization's infrastructure. Also, can you please get it reviewed by one more committer. I saw @guoma commenting on the JIRA, please check with him if he can do a second review.

arvindKandpal-ksolves commented:
Thanks @jyothsnakonisa. I will check with @smiklosovic regarding the CI run and request @Guoma for a second review on JIRA.

Maxwell-Guo commented on src/java/org/apache/cassandra/db/SystemKeyspace.java:
```diff
@@ -373,6 +373,7 @@ private SystemKeyspace()
                 + "keyspace_name text,"
                 + "rows_merged map<int, bigint>,"
                 + "compaction_properties frozen<map<text, text>>,"
+                + "compaction_type text,"
```
Hi , why you add this column for table compaction_history ? 
the column compaction_properties will also show the table's compaction type too.
we can see this jira  https://issues.apache.org/jira/browse/CASSANDRA-18061
The first time i add the column compaction_properties is used to show the type of  compaction task 
and the column type is text too, but we changed it to a frozen map after some discussion.

arvindKandpal-ksolves committed:
CASSANDRA-20081: Revert schema change, persist compaction type in properties map only

arvindKandpal-ksolves commented:
@Maxwell-Guo , I have updated the PR based on your feedback.
Reverted schema changes: Removed the new column.
Logic: compaction_type is now stored inside the compaction_properties map.

arvindKandpal-ksolves committed:
Fix conflicts and imports after rebase

Maxwell-Guo commented on src/java/org/apache/cassandra/db/compaction/CompactionHistoryTabularData.java:
```diff
@@ -86,6 +87,7 @@ public static TabularData from(UntypedResultSet resultSet) throws OpenDataExcept
             long bytesOut = row.getLong(ITEM_NAMES[5]);
             Map<Integer, Long> rowMerged = row.getMap(ITEM_NAMES[6], Int32Type.instance, LongType.instance);
             Map<String, String> compactionProperties = row.getMap(ITEM_NAMES[7], UTF8Type.instance, UTF8Type.instance);
+
```
nit: delete new added empty line.

Maxwell-Guo commented on src/java/org/apache/cassandra/tools/nodetool/stats/CompactionHistoryHolder.java:
```diff
@@ -90,6 +90,7 @@ private HashMap<String, Object> getAllAsMap(boolean humanReadable)
             compaction.put("bytes_out", FileUtils.stringifyFileSize(this.bytesOut, humanReadable));
             compaction.put("rows_merged", this.rowMerged);
             compaction.put("compaction_properties", this.compactionProperties);
+
```
nit: delete newly added empty line

Maxwell-Guo commented on test/unit/org/apache/cassandra/tools/nodetool/CompactionHistoryTest.java:
```diff
@@ -102,35 +102,59 @@ public void testCompactionProperties() throws Throwable
+
+        String cql = "select keyspace_name,columnfamily_name,compaction_properties from system." + SystemKeyspace.COMPACTION_HISTORY +
                      " where keyspace_name = '" + keyspace() + "' AND columnfamily_name = '" + currentTable() + "' ALLOW FILTERING";
+
```
delete newly added empty line.

Maxwell-Guo reviewed (approved):
some small nits, I can help with it when doing merge. will run ci latter.

arvindKandpal-ksolves commented:
Committed as 6d60102422d8409cdf8fabfb1c906afdf6a85859
